solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -0,0 +1,253 @@
+    }
+    // lookup leaf key for these children using path
+    // depending on the label, add to the parent at the right key/label
+    String trimmedPath = trimLastPound(cDocsPath);
+    // if the child doc's path does not end with #, it is an array(same string is returned by ChildDocTransformer#trimLastPound)
+    if (!parent.containsKey(trimmedPath) && (trimmedPath == cDocsPath)) {
+      List<SolrDocument> list = new ArrayList<>(children);
+      parent.setField(trimmedPath, list);
+      return;
+    }
+    // is single value
+    parent.setField(trimmedPath, ((List)children).get(0));
+  }
+
+  private static String getLastPath(String path) {
+    int lastIndexOfPathSepChar = path.lastIndexOf(PATH_SEP_CHAR);
+    if(lastIndexOfPathSepChar == -1) {
+      return path;
+    }
+    return path.substring(lastIndexOfPathSepChar + 1);
+  }
+
+  private static String trimLastPoundIfArray(String path) {
+    // remove index after last pound sign and if there is an array index e.g. toppings#1 -> toppings
+    // or return original string if child doc is not in an array ingredients# -> ingredients#
+    final int indexOfSepChar = path.lastIndexOf(NUM_SEP_CHAR);
+    if (indexOfSepChar == -1) {
+      return path;
+    }
+    int lastIndex = path.length() - 1;
+    boolean singleDocVal = indexOfSepChar == lastIndex;
+    return singleDocVal ? path: path.substring(0, indexOfSepChar);
+  }
+
+  private static String trimLastPound(String path) {
+    // remove index after last pound sign and index from e.g. toppings#1 -> toppings
+    int lastIndex = path.lastIndexOf('#');
+    return lastIndex == -1 ? path : path.substring(0, lastIndex);
+  }
+
+  /**
+   * Returns the *parent* path for this document.
+   * Children of the root will yield null.
+   */
+  private static String getParentPath(String currDocPath) {
+    // chop off leaf (after last '/')
+    // if child of leaf then return null (special value)
+    int lastPathIndex = currDocPath.lastIndexOf(PATH_SEP_CHAR);
+    return lastPathIndex == -1 ? null : currDocPath.substring(0, lastPathIndex);
+  }
+
+  /** Looks up the nest path.  If there is none, returns {@link #ANON_CHILD_KEY}. */
+  private static String getPathByDocId(int segDocId, SortedDocValues segPathDocValues) throws IOException {
+    int numToAdvance = segPathDocValues.docID() == -1 ? segDocId : segDocId - (segPathDocValues.docID());
+    assert numToAdvance >= 0;
+    boolean advanced = segPathDocValues.advanceExact(segDocId);
+    if (!advanced) {
+      return ANON_CHILD_KEY;
+    }
+    return segPathDocValues.binaryValue().utf8ToString();
+  }
+
+  /**
+   *
+   * @param segDocBaseId base docID of the segment
+   * @param RootId docID if the current root document
+   * @param lastDescendantId lowest docID of the root document's descendant
+   * @return the docID to loop and to not surpass limit of descendants to match specified by query
+   */
+  private int calcLimitIndex(int segDocBaseId, int RootId, int lastDescendantId) {
```

Comment:
The reason the sort is docID descending, is because this method initialises the index to the highest docID to match exhaust the limit, if possible, or the last descendant document.